Add documentation about C++20 modules support (#2105)
oneapi-src/oneTBB: doc/main/tbb_userguide/cxx20_modules_support.rst

.. note::
    To support C++20 modules, CMake* 3.28 or later and a compiler and generator with C++20 modules
    support are required. Refer to the
    `CMake* documentation <https://cmake.org/cmake/help/latest/manual/cmake-cxxmodules.7.html>`_
    for the list of supported compilers and generators.

.. code:: cmake

    cmake_minimum_required(VERSION 3.28)
    project(myapp CXX)

    set(CMAKE_CXX_STANDARD 20)

    find_package(TBB REQUIRED)

    add_executable(my_executable main.cpp)
    target_link_libraries(my_executable PRIVATE TBB::tbb)

    get_target_property(_tbb_include_dir TBB::tbb INTERFACE_INCLUDE_DIRECTORIES)
    target_sources(my_executable PRIVATE
        FILE_SET cxx_modules TYPE CXX_MODULES
        BASE_DIRS ${_tbb_include_dir}
        FILES ${_tbb_include_dir}/oneapi/tbb.cppm
    )

An then in your C++ source files, you can import the module:

.. literalinclude:: ./examples/cxx20_module_example.cpp
   :language: c++
   :prepend: import tbb;
   :start-after: /* begin_cxx20_modules_basic_example */
   :end-before: /* end_cxx20_modules_basic_example */

.. note::
    Translation units built with ``import tbb;`` are ABI-compatible with those built using

>     ``#include <oneapi/tbb.h>``.

Reviewer: ```suggestion
    Translation units that use ``import tbb;`` are ABI-compatible with those that use library headers.
```

.. _cxx20_modules_support:

C++20 Modules Support
=====================

.. note::
    Support for C++20 modules is experimental and subject to change in future releases.

|short_name| provides a C++20 module interface unit, ``tbb.cppm``, that lets you use |short_name| via

> ``import tbb;`` instead of ``#include <oneapi/tbb.h>``. The module is installed as a

Reviewer: I would not limit this by tbb.h only since it is applicable for any header.

```suggestion
``import tbb;`` instead of regular library headers. The module is installed as a
```

source file under ``<install-prefix>/include/oneapi/tbb.cppm`` and must be compiled as
part of your own build target.

CMake* Integration
******************

To add the ``tbb`` module to your CMake* target, locate ``tbb.cppm`` using the
``INTERFACE_INCLUDE_DIRECTORIES`` property of the ``TBB::tbb`` target and register it as
a ``CXX_MODULES`` file set:

.. note::

>     To support C++20 modules, CMake 3.28 or later and a compiler and generator with C++20 modules

Reviewer: ```suggestion
    To support C++20 modules, CMake* 3.28 or later and a compiler and generator with C++20 modules
```

    support are required. Refer to the

>     `CMake documentation <https://cmake.org/cmake/help/latest/manual/cmake-cxxmodules.7.html>`_

Reviewer: ```suggestion
    `CMake* documentation <https://cmake.org/cmake/help/latest/manual/cmake-cxxmodules.7.html>`_
```

    for the list of supported compilers and generators.

> .. code:: cmake

Reviewer 2: Consider checking code snippets on correctness automatically. That is making them to be built as part of CI. This will allow to catch divergence between the example and what's actually work.

Reviewer: As far as I remember, current doc examples testing only supports C++ snippets and their testing. 
Also, if we want to add a snippet that use modules, special support for modules should be added into doc-examples CI.

Author: Done.

    cmake_minimum_required(VERSION 3.28)
    project(myapp CXX)

    set(CMAKE_CXX_STANDARD 20)

    find_package(TBB REQUIRED)

    add_executable(my_executable main.cpp)
    target_link_libraries(my_executable PRIVATE TBB::tbb)

    get_target_property(_tbb_include_dir TBB::tbb INTERFACE_INCLUDE_DIRECTORIES)
    target_sources(my_executable PRIVATE
        FILE_SET cxx_modules TYPE CXX_MODULES
        BASE_DIRS ${_tbb_include_dir}
        FILES ${_tbb_include_dir}/oneapi/tbb.cppm
    )

An then in your C++ source files, you can import the module:

.. code:: cpp

    import tbb;

    int main() {
        tbb::parallel_for(0, 100, [](int i) { /* ... */ });
    }

.. note::
    Translation units built with ``import tbb;`` are ABI-compatible with those built using
    ``#include <oneapi/tbb.h>``. Because the exported API is attached to the global module
    fragment, the module name does not participate in symbol mangling, and both can be linked

>     into the same program.

Reviewer: Do we believe this implementation details are needed for a user? I feel that providing a binary compatibility guarantee should be enough.

Author: Removed implementation details.

> Usage Of Predefined Macros

Reviewer: I propose swapping this and the next section in the document. 

The reason is that the code snippet below defines the preview macro before `import tbb`. The reader that did not finish the page may assume it is enough to use the preview feature.

Author: Done.

**************************

C++20 modules do not export preprocessor macros. Macros defined in

> ``<oneapi/tbb/version.h>`` (such as version or feature-test macros) are **not** available

Reviewer: May be avoid using version.h only here since the macros can be defined in other headers as well?

```suggestion
|short_name| library headers (such as version or feature-test macros) are **not** available
```

> ``<oneapi/tbb/version.h>`` (such as ``TBB_VERSION`` or feature-test macros) are **not** available

Author: ```suggestion
``<oneapi/tbb/version.h>`` (such as version or feature-test macros) are **not** available
```

after ``import tbb;``.

> As a workaround, include the version header alongside the module import.

Reviewer: ```suggestion
As a workaround, include the `<oneapi/tbb/version.h>` header alongside the module import.
```

.. code:: cpp

    #define TBB_PREVIEW_FEATURE_X 1
    #include <oneapi/tbb/version.h>  // macros available here
    import tbb;

    static_assert(TBB_VERSION_MAJOR >= 2023, "Major version 2023 or later required");

    // Feature-test macros are also available
    #ifdef TBB_HAS_FEATURE_X

>     // use feature X

Reviewer: After swapping the sections, I would add another explicit caution here that if the module is built without TBB_PREVIEW_FEATURE_X defined, even if TBB_HAS_FEATURE_X is 1 in this code snippet, the feature is not available. Or something in this manner.

Author: Added a caution and updated the example with assumption of proposed CMake integration usage.

    #endif

Enabling Preview Features
*************************

> |short_name| preview features are gated by ``TBB_PREVIEW_*`` macros that are normally defined
> before including TBB headers. With modules, macros defined by the consumer before

Reviewer 2: Small clarification:
```suggestion
|short_name| preview features are gated by ``TBB_PREVIEW_*`` macros. Each preview macro
needs to be defined before a corresponding TBB header declaring the preview feature is
included. With modules, macros defined by the consumer before
```

Author: > needs to be defined before a corresponding TBB header

That is not true. Preview macros should be defined before inclusion of any TBB header as many of them include `_config.h`.

``import tbb;`` **cannot** affect the module's already-compiled interface.

To enable a preview feature, define the corresponding macro when **compiling**
``tbb.cppm`` itself. In CMake*, use ``target_compile_definitions``:

.. code:: cmake

    find_package(TBB REQUIRED)

    add_executable(my_executable main.cpp)
    target_link_libraries(my_executable PRIVATE TBB::tbb)

    get_target_property(_tbb_include_dir TBB::tbb INTERFACE_INCLUDE_DIRECTORIES)
    target_sources(my_executable PRIVATE
        FILE_SET cxx_modules TYPE CXX_MODULES
        BASE_DIRS ${_tbb_include_dir}
        FILES ${_tbb_include_dir}/oneapi/tbb.cppm
    )

>     # Enable preview features when compiling the module

Reviewer 2: ```suggestion
    # Enable preview feature X when compiling the module
```
